src/Utilities/EmailSender.cs

```diff
@@ -1,50 +1,69 @@
+        private string GreetingMessageSubject { get; } = "Welcome to ePiggy!";
+        private string GreetingMessageBody { get; } = "Thank you for joining ePiggy! Your wallet thanks you as well!";
+        private string FarewellMessageSubject { get; } = "Farewell from the ePiggy team";
+        private string FarewellMessageBody { get; } = "We're sorry to see you go, please leave your feedback of our app.";
+        private int MaxRandValue { get; }
+        private string SmtpEmail { get; }
+        private int Port { get; }
 
-        public static int SendRecoveryCode(string email)
+        public EmailSender(int maxRandValue = 999999, string smtpEmail = "smtp.gmail.com", int port = 587)
         {
-            using var db = new DatabaseContext();
-            var user = db.Users.FirstOrDefault(a => a.Email == email);
-
-            if (user == null)
-            {
-                return 0;
-            }
+            MaxRandValue = maxRandValue;
+            SmtpEmail = smtpEmail;
+            Port = port;
+        }
 
+        public async Task<int> SendRecoveryCodeAsync(string email)
+        {
             var rand = new Random();
             var randomCode = rand.Next(MaxRandValue);
+            var message = new MailMessage(OurEmail, email, RecoveryMessageSubject, RecoveryMessageBody + randomCode);
+            await SendAsync(message);
+            return randomCode;
+        }
 
-            var message = new MailMessage();
-            message.To.Add(email);
-            message.From = new MailAddress(EmailAddress);
-            message.Body = RecoveryMessageBody + randomCode;
-            message.Subject = RecoveryMessageSubject;
+        public async Task SendFarewellEmailAsync(object sender, int id)
+        {
+            var user = UserDatabase.GetUser(id);
+            if (user is null) return;
+            await SendFarewellEmailAsync(user.Email);
+        }
 
-            var smtp = new SmtpClient(SmtpEmail)
+        public async Task SendFarewellEmailAsync(string email)
+        {
+            var message = new MailMessage(OurEmail, email, FarewellMessageSubject, FarewellMessageBody);
+            await SendAsync(message);
+        }
+
+        public async Task SendGreetingEmailAsync(string email)
+        {
+            var message = new MailMessage(OurEmail, email, GreetingMessageSubject, GreetingMessageBody);
+            await SendAsync(message);
+        }
+
+        private async Task SendAsync(MailMessage message)
+        {
+            var smtp = new SmtpClient(SmtpEmail, Port)
             {
                 EnableSsl = true,
-                Port = PortConst,
                 DeliveryMethod = SmtpDeliveryMethod.Network,
-                Credentials = new NetworkCredential(EmailAddress, EmailPassword)
+                Credentials = new NetworkCredential(OurEmail, EmailPassword)
             };
-            smtp.Send(message);
-
-            return randomCode;
+            await smtp.SendMailAsync(message);
```

Comment:
Just curious, did these await everywhere improve the speed? I understand that it protects sender from garbage collection, but what about performance?

Reply:
I don't think it has impacted performance very much.